test/session.test.js

```diff
@@ -0,0 +1,160 @@
+    return application.storageService.getValue(StorageKey.Session);
+  }
+
+  it('should fail when a sync request is perfomed with an expired access token', async function () {
+    const currentSession = this.application.apiService.session;
+
+    const delayBeforeNextRequest = getDelayBeforeNextRequest(currentSession);
+    await Factory.sleep(delayBeforeNextRequest);
+
+    const response = await this.application.apiService.sync([]);
+
+    expect(response).to.have.property('status');
+    expect(response.status).to.equal(498);
+
+    expect(response).to.have.property('error');
+
+    expect(response.error).to.have.property('tag');
+    expect(response.error.tag).to.equal('expired-access-token');
+
+    expect(response.error).to.have.property('message');
+    expect(response.error.message).to.equal('The provided access token has expired.');
+  }).timeout(10000);
+
+  it('should return the new session in the response when refreshed', async function () {
+    const response = await this.application.apiService.refreshSession();
+
+    expect(response).to.have.property('status');
+    expect(response.status).to.equal(200);
+
+    expect(response).to.have.property('token');
+    expect(response.token).to.be.a('string');
+    expect(response.token).to.not.be.empty;
+
+    expect(response).to.have.property('session');
+
+    expect(response.session).to.have.property('expire_at');
+    expect(response.session.expire_at).to.be.a('number');
+
+    expect(response.session).to.have.property('refresh_token');
+    expect(response.session.refresh_token).to.not.be.empty;
+  }).timeout(10000);
+
+  it('should be refreshed if access token is expired', async function () {
+    // Saving the current session information for later...
+    const sessionBeforeSync = this.application.apiService.session;
+
+    // Waiting enough time for the access token to expire, before performing a new sync request.
+    const delayBeforeNextRequest = getDelayBeforeNextRequest(sessionBeforeSync);
+    await Factory.sleep(delayBeforeNextRequest);
+
+    // Performing a sync request with an expired access token.
+    await this.application.sync(syncOptions);
+
+    // After the above sync request is completed, we obtain the session information.
+    const sessionAfterSync = this.application.apiService.session;
+
+    expect(sessionBeforeSync).to.not.equal(sessionAfterSync);
+    expect(sessionBeforeSync.accessToken).to.not.equal(sessionAfterSync.accessToken);
+    expect(sessionBeforeSync.refreshToken).to.not.equal(sessionAfterSync.refreshToken);
+    expect(sessionBeforeSync.expireAt).to.be.lessThan(sessionAfterSync.expireAt);
+  }).timeout(20000);
+
+  it('should be consistent between storage and apiService', async function () {
+    const sessionFromStorage = await getSessionFromStorage(this.application);
+    const sessionFromApiService = this.application.apiService.session;
+
+    expect(sessionFromStorage).to.equal(sessionFromApiService);
```

Comment:
How is it possible that a raw session from storage is equal to a Session object 🤔The former should be a generic JS object, and the latter would be a Session instance.

Reply:
Good question. I sort of missed that detail there. Somehow, the session from `storageService` is a `Session` object. But it is not clear to my why...